docs/architecture/adr-006-upgrade-client-implementation.md

```diff
@@ -0,0 +1,433 @@
+      upgraded_consensus_state: Any,
+      proof_upgrade_client: MerkleProof,
+      proof_upgrade_consensus_state: MerkleProof,
+      root: &CommitmentRoot,
+   ) -> Result<(), ClientError>;
+   ```
+
+* And `update_state_with_upgrade_client` method
+
+   ```rust
+   fn update_state_with_upgrade_client(
+      &self,
+      upgraded_client_state: Any,
+      upgraded_consensus_state: Any,
+   ) -> Result<UpdatedState, ClientError>;
+   ```
+
+There is also a need to move away from `upgrade` method of `ClientState` trait
+leaving only zeroed out fields through `zero_custom_fields()`. New state
+commitment should be done in `update_state_with_upgrade_client` and stored via
+`store_client_result` of keeper context, so the upgrade part is no longer
+necessary.
+
+Listed below are the code snippets that correspond to the third step of the
+previous section as mentioned:
+
+* **Basic Validations**
+
+1. ```rust
+   if old_client_state.is_frozen() {
+      return Err(ContextError::ClientError(ClientError::ClientFrozen {
+         client_id,
+      }));
+   }
+   ```
+
+2. ```rust
+   let old_consensus_state = ctx
+      .consensus_state(&client_id, &old_client_state.latest_height())
+      .map_err(|_| ClientError::ConsensusStateNotFound {
+         client_id: client_id.clone(),
+         height: old_client_state.latest_height(),
+      })?;
+
+   let now = ctx.host_timestamp()?;
+   let duration = now
+      .duration_since(&old_consensus_state.timestamp())
+      .ok_or_else(|| ClientError::InvalidConsensusStateTimestamp {
+         time1: old_consensus_state.timestamp(),
+         time2: now,
+      })?;
+
+   if old_client_state.expired(duration) {
+      return Err(ClientError::HeaderNotWithinTrustPeriod {
+         latest_time: old_consensus_state.timestamp(),
+         update_time: now,
+      });
+   };
+   ```
+
+3. ```rust
+   // Decode the proto message into a domain message inside deliver() function
+   // include the proofs in the message
+   let envelope = decode(message)?;
+   ```
+
+* **Chain-specific Validations**
+
+4. ```rust
+   let upgraded_tm_client_state = TmClientState::try_from(upgraded_client_state)?;
+   ```
+
+5. ```rust
+   let upgraded_tm_cons_state = TmConsensusState::try_from(upgraded_consensus_state)?;
+   ```
+
+6. ```rust
+   if self.latest_height() >= upgraded_tm_client_state.latest_height() {
+      return Err(ClientError::LowUpgradeHeight {
+            upgraded_height: self.latest_height(),
+            client_height: upgraded_tm_client_state.latest_height(),
+      });
+   }
+   ```
+
+7. ```rust
+   // Ensure that the new unbonding period is not shorter than the current's client unbonding period
+   if self.unbonding_period > upgraded_tm_client_state.unbonding_period {
+      return Err(ClientError::ClientSpecific {
+            description:
+               "cannot upgrade client with a shorter unbonding period than the current one"
+                  .to_string(),
+      });
+   }
```

Comment:
I don't see this in ibc-go. I'd be interested to know why they don't have that check, which would inform whether or not we should keep it.

Same goes for any other check that ibc-go doesn't have.

Reply:
Per our discussion, we'll keep this check, unless there's a good reason or serious need for shorter unbounding periods